### ISSUE TITLE
Local function with decreases clause leads to ASSERTION FAILURE

### ISSUE THREAD

**Issue by markulf** [FStarLang/FStar#1361]
Just following instructions in a strange error message produced by making a top-level function local.

```
ASSERTION FAILURE: Impossible: missing universe instantiation on length
F* may be in an inconsistent state.
Please file a bug report, ideally with a minimized version of the program 
that triggered the error.
```

```
module LocalFunctionBug

open FStar.List.Tot

let rec f (is:list int)
: Tot(list int)
           (decreases %[List.length is]) 
  =
match is with
| [] -> []
| i::is -> []

let f1 (is:list int) = f is 

let f2 (is:list int) = 
  let rec localf (is:list int)
    : Tot(list int)
           (decreases %[List.length is]) 
      =
    match is with
    | [] -> []
    | i::is -> []
  in 
    localf is 
```

**Comment by aseemr** [comment]
Seems like a bug in type checking the decreases clauses of inner let recs. Explicitly instantiating the universe  (`decreases (length u#0 is)`) would work as a work around.

**Comment by markulf** [comment]
thanks, I will add the bug repro to regression

**Comment by mtzguido** [comment]
This seems to work now in master. There was one fix regarding decreases clauses (#1360, interestingly enough), but unsure that was the actual bullet. @markulf, if you have more examples of this bug would you like to try them out, or should we close this?

**Comment by markulf** [comment]
thanks for fixing this!
